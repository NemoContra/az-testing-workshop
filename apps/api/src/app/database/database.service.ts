import { Injectable } from '@nestjs/common';
import { Contract } from '../models/contract.model';

@Injectable()
export class DatabaseService {
  private mockData: Contract[] = [
    {
      id: '123456789',
      contractNumber: '1/2345678/9',
      start: '2024-01-01',
      premium: 42.42,
      person: { firstname: 'Homer', lastname: 'Simpson', dateOfBirth: '1961-05-16' },
    },
    {
      id: '123456788',
      contractNumber: '1/2345678/8',
      start: '2024-02-01',
      premium: 99.000,
      person: { firstname: 'Bart', lastname: 'Simpson', dateOfBirth: '1995-08-21' },
    },
  ];

  public getAllContracts(): Promise<Contract[]> {
    return Promise.resolve(this.mockData);
  }

  public getContractById(contractId: string): Promise<Contract | undefined> {
    return Promise.resolve(this.mockData.find(({ id }) => id === contractId));
  }

  public updateContract(updatedContract: Partial<Contract>): Promise<Contract | undefined> {
    const existingContract = this.mockData.find(({ id }) => id === updatedContract.id);

    if (!existingContract) {
      return Promise.resolve(undefined);
    }

    const updatedContracts = this.mockData.filter(({ id }) => id !== updatedContract.id);

    this.mockData = [...updatedContracts, {...existingContract, ...updatedContract}];

    return Promise.resolve({...existingContract, ...updatedContract});
  }
}
