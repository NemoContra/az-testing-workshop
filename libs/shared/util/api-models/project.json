{
  "name": "shared-util-api-models",
  "$schema": "../../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util/api-models/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint"
    }
  },
  "tags": ["scope:shared", "type:util"]
}
