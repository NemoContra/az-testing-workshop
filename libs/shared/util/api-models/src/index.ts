export * from './lib/shared-util-api-models';
