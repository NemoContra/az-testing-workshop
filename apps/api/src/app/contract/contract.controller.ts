import {
  Body,
  Controller,
  Get,
  NotFoundException,
  Param,
  Put, Query,
} from '@nestjs/common';
import { DatabaseService } from '../database/database.service';
import { Contract } from '@az-testing-workshop/shared/util/api-models';

@Controller('contracts')
export class ContractController {
  constructor(private readonly databaseService: DatabaseService) {}

  @Get('')
  getAllContracts(@Query('query') searchQuery?: string): Promise<Contract[]> {
    return this.databaseService.getAllContracts(searchQuery);
  }

  @Get(':id')
  async getContractById(
    @Param('id') id: string
  ): Promise<Contract | undefined> {
    const contract = await this.databaseService.getContractById(id);

    if (!contract) {
      throw new NotFoundException();
    }

    return contract;
  }

  @Put('')
  async updateContract(@Body() contract: Contract): Promise<Contract> {
    const updatedContract = await this.databaseService.updateContract(contract);

    if (!updatedContract) {
      throw new NotFoundException();
    }

    return updatedContract;
  }
}
