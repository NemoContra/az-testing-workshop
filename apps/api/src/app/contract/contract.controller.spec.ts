import { Test, TestingModule } from '@nestjs/testing';
import { ContractController } from './contract.controller';
import { DatabaseService } from '../database/database.service';
import { NestApplication } from '@nestjs/core';
import request from 'supertest';
import { Contract } from '@az-testing-workshop/shared/util/api-models';

describe('ContractController', () => {
  let app: NestApplication;
  let databaseService: DatabaseService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [ContractController],
      providers: [DatabaseService],
    }).compile();

    app = module.createNestApplication();
    databaseService = app.get(DatabaseService);
    await app.init();
  });

  afterAll(() => app.close());

  describe('GET getAllContracts', () => {
    it('should return all contracts', () => {
      jest.spyOn(databaseService, 'getAllContracts').mockResolvedValue([
        {
          id: '123456789',
          contractNumber: '1/2345678/9',
          start: '2024-01-01',
          premium: 42.42,
          person: {
            firstname: 'Homer',
            lastname: 'Simpson',
            dateOfBirth: '1961-05-16',
          },
        },
      ] satisfies Contract[]);

      return request(app.getHttpServer())
        .get('/contracts')
        .expect(200)
        .then(({ body }) => {
          expect(app.get(DatabaseService).getAllContracts).toHaveBeenCalled();
          expect(body).toMatchSnapshot();
        });
    });
  });

  describe('GET getContractById', () => {
    it('should return contract by id', () => {
      jest.spyOn(databaseService, 'getContractById').mockResolvedValue({
        id: '123456789',
        contractNumber: '1/2345678/9',
        start: '2024-01-01',
        premium: 42.42,
        person: {
          firstname: 'Homer',
          lastname: 'Simpson',
          dateOfBirth: '1961-05-16',
        },
      } satisfies Contract);

      return request(app.getHttpServer())
        .get('/contracts/123456789')
        .expect(200)
        .then(({ body }) => {
          expect(app.get(DatabaseService).getContractById).toHaveBeenCalledWith(
            '123456789'
          );
          expect(body).toMatchSnapshot();
        });
    });

    it('should return 404 not found', () => {
      jest
        .spyOn(databaseService, 'getContractById')
        .mockResolvedValue(undefined);

      return request(app.getHttpServer())
        .get('/contracts/123')
        .expect(404)
        .then(({ body }) => {
          expect(app.get(DatabaseService).getContractById).toHaveBeenCalledWith(
            '123'
          );
          expect(body).toMatchSnapshot();
        });
    });
  });

  describe('PUT updateContract', () => {
    it('should update a contract and return the new one', () => {
      jest.spyOn(databaseService, 'updateContract').mockResolvedValue({
        id: '123456789',
        contractNumber: '1/2345678/9',
        start: '2024-01-01',
        premium: 5000,
        person: {
          firstname: 'Homer',
          lastname: 'Stark',
          dateOfBirth: '1961-05-16',
        },
      } satisfies Contract);

      return request(app.getHttpServer())
        .put('/contracts')
        .send({ id: '123456789', premium: 5000 })
        .expect(200)
        .then(({ body }) => {
          expect(app.get(DatabaseService).updateContract).toHaveBeenCalledWith({
            id: '123456789',
            premium: 5000,
          } satisfies Partial<Contract>);
          expect(body).toMatchSnapshot();
        });
    });

    it('should return 404 not found', () => {
      jest
        .spyOn(databaseService, 'updateContract')
        .mockResolvedValue(undefined);

      return request(app.getHttpServer())
        .put('/contracts')
        .send({ id: '123', premium: 5000 })
        .expect(404)
        .then(({ body }) => {
          expect(app.get(DatabaseService).updateContract).toHaveBeenCalledWith({
            id: '123',
            premium: 5000,
          } satisfies Partial<Contract>);
          expect(body).toMatchSnapshot();
        });
    });
  });
});
