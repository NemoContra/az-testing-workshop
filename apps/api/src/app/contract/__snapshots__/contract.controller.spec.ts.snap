// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractController GET getAllContracts should return all contracts 1`] = `
[
  {
    "contractNumber": "1/2345678/9",
    "id": "123456789",
    "person": {
      "dateOfBirth": "1961-05-16",
      "firstname": "Homer",
      "lastname": "Simpson",
    },
    "premium": 42.42,
    "start": "2024-01-01",
  },
]
`;

exports[`ContractController GET getContractById should return 404 not found 1`] = `
{
  "message": "Not Found",
  "statusCode": 404,
}
`;

exports[`ContractController GET getContractById should return contract by id 1`] = `
{
  "contractNumber": "1/2345678/9",
  "id": "123456789",
  "person": {
    "dateOfBirth": "1961-05-16",
    "firstname": "Homer",
    "lastname": "Simpson",
  },
  "premium": 42.42,
  "start": "2024-01-01",
}
`;

exports[`ContractController PUT updateContract should return 404 not found 1`] = `
{
  "message": "Not Found",
  "statusCode": 404,
}
`;

exports[`ContractController PUT updateContract should update a contract and return the new one 1`] = `
{
  "contractNumber": "1/2345678/9",
  "id": "123456789",
  "person": {
    "dateOfBirth": "1961-05-16",
    "firstname": "Homer",
    "lastname": "Stark",
  },
  "premium": 5000,
  "start": "2024-01-01",
}
`;
